File: app/src/main/java/com/bridou_n/beaconscanner/utils/BluetoothManager.java
package com.bridou_n.beaconscanner.utils;

import android.bluetooth.BluetoothAdapter;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.support.annotation.Nullable;

import com.bridou_n.beaconscanner.events.Events;

import javax.inject.Inject;

import rx.Observable;
import rx.subjects.BehaviorSubject;

public class BluetoothManager {

    private final BluetoothAdapter adapter;

    private final BehaviorSubject<Object> subject;

    @Inject
    public BluetoothManager(@Nullable BluetoothAdapter adapter, Context context) {
        this.adapter = adapter;
        this.subject = BehaviorSubject.create(new Events.BluetoothState(adapter != null ? adapter.getState() : BluetoothAdapter.STATE_OFF));

        BroadcastReceiver receiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                if (BluetoothAdapter.ACTION_STATE_CHANGED.equals(intent.getAction())) {
                    subject.onNext(new Events.BluetoothState(
                            intent.getIntExtra(BluetoothAdapter.EXTRA_STATE,
                                    BluetoothAdapter.ERROR)));
                }
            }
        };
        context.registerReceiver(receiver, new IntentFilter(BluetoothAdapter.ACTION_STATE_CHANGED));
    }

    public void disable() {
        if (adapter != null) {
            adapter.disable();
        }
    }

    public void enable() {
        if (adapter!= null) {
            adapter.enable();
        }
    }

    public boolean isEnabled() {
        return (adapter != null) && adapter.isEnabled();
    }

    public Observable<Object> observe() {
        return subject;
    }

    public void toggle() {
        if (isEnabled()) {
            disable();
        } else {
            enable();
        }
    }
}

File: app/src/main/java/com/bridou_n/beaconscanner/utils/BluetoothManager.kt
package com.bridou_n.beaconscanner.utils

import android.bluetooth.BluetoothAdapter
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import com.bridou_n.beaconscanner.features.beaconList.BeaconListActivity
import io.reactivex.Flowable
import io.reactivex.processors.BehaviorProcessor
import javax.inject.Inject

class BluetoothManager @Inject constructor(private val adapter: BluetoothAdapter?, context: Context) {

    private val subject: BehaviorProcessor<BeaconListActivity.BluetoothState> =
            BehaviorProcessor.createDefault<BeaconListActivity.BluetoothState>(getStaeFromAdapterState(adapter?.state ?: BluetoothAdapter.STATE_OFF))

    init {
        val receiver = object : BroadcastReceiver() {
            override fun onReceive(ctx: Context, intent: Intent) {
                if (BluetoothAdapter.ACTION_STATE_CHANGED == intent.action) {
                    val state = getStaeFromAdapterState(intent.getIntExtra(BluetoothAdapter.EXTRA_STATE, BluetoothAdapter.ERROR))

                    subject.onNext(state)
                }
            }
        }
        context.registerReceiver(receiver, IntentFilter(BluetoothAdapter.ACTION_STATE_CHANGED))
    }

    fun getStaeFromAdapterState(state: Int) : BeaconListActivity.BluetoothState {
        return when (state) {
            BluetoothAdapter.STATE_OFF -> BeaconListActivity.BluetoothState.STATE_OFF
            BluetoothAdapter.STATE_TURNING_OFF -> BeaconListActivity.BluetoothState.STATE_TURNING_OFF
            BluetoothAdapter.STATE_ON -> BeaconListActivity.BluetoothState.STATE_ON
            BluetoothAdapter.STATE_TURNING_ON -> BeaconListActivity.BluetoothState.STATE_TURNING_ON
            else -> BeaconListActivity.BluetoothState.STATE_OFF
        }
    }

    fun disable() = adapter?.disable()

    fun enable() = adapter?.enable()

    fun asFlowable(): Flowable<BeaconListActivity.BluetoothState> {
        return subject
    }

    fun isEnabled() = adapter?.isEnabled == true

    fun toggle() {
        if (isEnabled()) {
            disable()
        } else {
            enable()
        }
    }
}
